troupe.go

```diff
@@ -65,66 +68,102 @@ func NewTroupe(cfg Config) (*Troupe, error) {
 		minActors:          cfg.Min,
 		maxActors:          cfg.Max,
 		defaultActorConfig: bCfg,
+		r:                  rand.New(rand.NewSource(time.Now().UnixNano())),
 	}, nil
 }
 
 // Shutdown shuts down the Troupe
-func (s *Troupe) Shutdown() {
-	s.ActorMutex.Lock()
-	s.shutdown = true
-	for _, b := range s.Actors {
-		b.stop()
+func (t *Troupe) Shutdown() error {
+	t.ActorMutex.Lock()
+	defer t.ActorMutex.Unlock()
+	if t.shutdown {
+		return ShuttingDownError("you cannot call shutdown more than once on a troupe")
+	}
+	t.shutdown = true
+	// Stop all of them right away, to shut off their ability to accept work
+	// Is this necessary to break into 2 steps?
+	for _, a := range t.Actors {
+		a.stop()
+	}
+
+	fmt.Printf("Number of goroutines: %d\n", runtime.NumGoroutine())
+	return nil
+}
+
+// Join is something i'm experimenting with to call after Shutdown, so you know when all work has ceased
+func (t *Troupe) Join() {
+	t.ActorMutex.Lock()
+	// wait until they all finish their backlogs of work.
+	for _, a := range t.Actors {
+		a.join()
 	}
-	s.ActorMutex.Unlock()
+	t.ActorMutex.Unlock()
+	fmt.Printf("Number of goroutines: %d\n", runtime.NumGoroutine())
 }
 
 // Assign will distribute a Letter to the first available Actor. If there are no available Actors (that is, no Actors
 // currently free from work, it grow the pool of Actors by 1. If the pool is already full, it will assign the work to the
 // Actor who has least-recently been assigned work. If all Actors have their mailboxes full, Assign will block until
 // one becomes free.
-func (s *Troupe) Assign(w Work) error {
-	s.ActorMutex.Lock()
-	defer s.ActorMutex.Unlock()
-	if s.shutdown {
-		return errors.New("Unable to assign work - shutting down")
+// Testing this vs AssignRand atm
+func (t *Troupe) Assign(w Work) error {
+	t.ActorMutex.Lock()
+	defer t.ActorMutex.Unlock()
+	if t.shutdown {
+		return ShuttingDownError("unable to assign work, shutting down")
 	}
 	var item *Actor
 	// First, do a best-effort attempt to find any Actors currently doing 0 work
-	// This will make sure the assigned letter is handled more quickly.
-	for i, b := range s.Actors {
-		if !b.IsBusy() {
+	// This will make sure the assigned work is handled more quickly.
+	for i, a := range t.Actors {
+		if !a.IsBusy() {
 			// Remove it, assign to it, push it back on the stack
-			item = s.Actors[i]
-			item.Accept(w)
+			item = t.Actors[i]
+			if err := item.Accept(w); err != nil {
+				return err
+			}
 			// Remove the item from where we found it
-			s.Actors = append(s.Actors[:i], s.Actors[i+1:]...)
+			t.Actors = append(t.Actors[:i], t.Actors[i+1:]...)
```

Comment:
Potentially, yes - that's where you'd wanna do some ingress measurement alongside latency measurements, and determine how many concurrent workers you need - even so far as spinning up multiple consumer/worker processes on multiple boxes to keep up.

The issue of messages sitting in flight, not processed, and lost if the process dies is definitely an issue. My initial assumption is that you'd need to use a somewhat durable system to ingest messages this way (kafka, sqs, etc).